### DISCUSSION TITLE
How to completely clean up and remove vpn from network stack?

### DISCUSSION THREAD

**Discussion by richm** [libreswan/libreswan#1738] [General]
Hello (hi @letoams !) - I'm working on the Ansible linux system role for vpn (libreswan) - https://github.com/linux-system-roles/vpn/
We run integration tests in conjunction with other system roles.  One of the vpn tests - https://github.com/linux-system-roles/vpn/blob/main/tests/tests_subnet_to_subnet.yml - leaves the network stack on the machine in a strange state that causes failures in some of the network role tests - specifically, tests that use dnsmasq to assign IPv4 addresses via dhcp.  I've tried centos stream 9, not sure if it has the same problem on other platforms.  I thought that doing this cleanup would suffice - https://github.com/linux-system-roles/vpn/blob/main/tests/tasks/cleanup.yml - but I can see that this leaves the `ip_vti0` network interface as well as several kernel modules.
I found a script in the libreswan source code called _stackmanager which looks like the `stop_xfrm` function does the cleanup.  This does remove the `ip_vti0` interface as well as most of the kernel modules, but the network stack is still in a strange state.  I'm wondering if there is some way to completely wipe out whatever libreswan does to the network stack to restore it to the previous state?

**Comment by letoams** [comment]
On Wed, 19 Jun 2024, Richard Megginson wrote:

> Hello (hi @letoams !) - I'm working on the Ansible linux system role for vpn (libreswan) -
> https://github.com/linux-system-roles/vpn/
> We run integration tests in conjunction with other system roles. One of the vpn tests -
> https://github.com/linux-system-roles/vpn/blob/main/tests/tests_subnet_to_subnet.yml - leaves the network stack
> on the machine in a strange state that causes failures in some of the network role tests - specifically, tests
> that use dnsmasq to assign IPv4 addresses via dhcp. I've tried centos stream 9, not sure if it has the same
> problem on other platforms. I thought that doing this cleanup would suffice -
> https://github.com/linux-system-roles/vpn/blob/main/tests/tasks/cleanup.yml - but I can see that this leaves
> the ip_vti0 network interface as well as several kernel modules.

Leaving kernel modules loaded should not cause failure. You should
investigate what the actual failure problem is.

Note that libreswan 5.x no longer uses _stackmanager, and assumes all
modules are automatically loaded or preloaded by the system.

The ip_vti0 network interface is an artifact of loading ip_vti, nothing
less, nothing more.

> I found a script in the libreswan source code called _stackmanager which looks like the stop_xfrm function does
> the cleanup. This does remove the ip_vti0 interface as well as most of the kernel modules, but the network
> stack is still in a strange state. I'm wondering if there is some way to completely wipe out whatever libreswan
> does to the network stack to restore it to the previous state?

libreswan 5.0+ no longer cleans up in the ipsec.service file or
otherwise. During a regular shutdown it will delete all connections,
which should clean up all xfrm kernel state and policies, but if there
are bugs there might be left over state. You can check with "ip xfrm
state" and "ip xfrm policy" (but aren't you destroying the network
namespace anyway?).

I think you also might be using the ipsec whack --shutdown --leave-state
option, which explicitely DOES NOT clean up anything. This was so
packets could flow while the pod was restarting.

libreswan DOES clean up all state on startup.

Paul

**Reply by richm** [reply]
Thanks.

Turns out this was a firewall issue - https://github.com/linux-system-roles/vpn/pull/158
